component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java

```diff
@@ -475,4 +502,184 @@ private boolean isValidTokenLocalUser(int token, String username, Authentication
+
+		// Return if account lock handler is not enabled.
+		for (Property connectorConfig : connectorConfigs) {
+			if ((TOTPAuthenticatorConstants.PROPERTY_ACCOUNT_LOCK_ON_FAILURE.equals(connectorConfig.getName())) &&
+					!Boolean.parseBoolean(connectorConfig.getValue())) {
+				return;
+			}
+		}
+
+		String username = context.getProperty("username").toString();
+		String usernameWithDomain = IdentityUtil.addDomainToName(authenticatedUser.getUserName(),
+				authenticatedUser.getUserStoreDomain());
+		try {
+			UserRealm userRealm = TOTPUtil.getUserRealm(username);
+			UserStoreManager userStoreManager = userRealm.getUserStoreManager();
+
+			// Avoid updating the claims if they are already zero.
+			String[] claimsToCheck = {TOTPAuthenticatorConstants.TOTP_FAILED_ATTEMPTS_CLAIM,
+					TOTPAuthenticatorConstants.FAILED_LOGIN_LOCKOUT_COUNT_CLAIM};
+			Map<String, String> userClaims = userStoreManager.getUserClaimValues(usernameWithDomain, claimsToCheck,
+					UserCoreConstants.DEFAULT_PROFILE);
+			String failedSmsOtpAttempts = userClaims.get(TOTPAuthenticatorConstants.TOTP_FAILED_ATTEMPTS_CLAIM);
+			String failedLoginLockoutCount =
+					userClaims.get(TOTPAuthenticatorConstants.FAILED_LOGIN_LOCKOUT_COUNT_CLAIM);
+
+			if (NumberUtils.isNumber(failedSmsOtpAttempts) && Integer.parseInt(failedSmsOtpAttempts) > 0 ||
+					NumberUtils.isNumber(failedLoginLockoutCount) && Integer.parseInt(failedLoginLockoutCount) > 0) {
+				Map<String, String> updatedClaims = new HashMap<>();
+				updatedClaims.put(TOTPAuthenticatorConstants.TOTP_FAILED_ATTEMPTS_CLAIM, "0");
+				updatedClaims.put(TOTPAuthenticatorConstants.FAILED_LOGIN_LOCKOUT_COUNT_CLAIM, "0");
+				userStoreManager
+						.setUserClaimValues(usernameWithDomain, updatedClaims, UserCoreConstants.DEFAULT_PROFILE);
+			}
+		} catch (UserStoreException e) {
+			log.error("Error while resetting failed TOTP attempts", e);
+			String errorMessage =
+					String.format("Failed to reset failed attempts count for user : %s.",
+							authenticatedUser.getUserName());
+			throw new AuthenticationFailedException(errorMessage, e);
+		}
+	}
+
+	private Map<String, String> getUserClaimValues(AuthenticatedUser authenticatedUser, String username)
+			throws AuthenticationFailedException {
+
+		Map<String, String> claimValues;
+		try {
+			UserRealm userRealm = TOTPUtil.getUserRealm(username);
+			UserStoreManager userStoreManager = userRealm.getUserStoreManager();
+			claimValues = userStoreManager.getUserClaimValues(IdentityUtil.addDomainToName(
+					authenticatedUser.getUserName(), authenticatedUser.getUserStoreDomain()), new String[]{
+							TOTPAuthenticatorConstants.TOTP_FAILED_ATTEMPTS_CLAIM,
+							TOTPAuthenticatorConstants.FAILED_LOGIN_LOCKOUT_COUNT_CLAIM},
+					UserCoreConstants.DEFAULT_PROFILE);
+		} catch (UserStoreException e) {
+			log.error("Error while reading user claims of user: " + authenticatedUser.getUserName(), e);
```

Comment:
Usually, we do not log an error and throw an exception because this will add duplicate entries in the logs. Better to add a debug log. Shall we fix this in other places as well?

Reply:
Fixed in 8307785